### ISSUE TITLE
[BUG] Multiple SQLAlchemy engines are created for a single DB backend store

### ISSUE THREAD

**Issue by alparslanavci** [mlflow/mlflow#12982]
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Other

### Willingness to contribute

No. I cannot contribute a bug fix at this time.

### MLflow version

- Tracking server: 2.14.2

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 
- **Python version**: 3.11

### Describe the problem

When starting the Tracking Server, multiple SQLAlchemy engines are created, all connecting to the same PostgreSQL backend store. This is evident from the logs, which display several entries related to engine creation. Additionally, the number of active database connections in PostgreSQL matches the number of SQLAlchemy engines being created.

This issue was previously reported in #3203 and was expected to be resolved by #4663. However, the problem persists in version 2.14.2.

By the way, we tried setting the SQLAlchemy options mentioned [here](https://mlflow.org/docs/2.14.3/tracking/backend-stores.html#sqlalchemy-options), but it didn't help. Although these options are applied to each engine, since multiple engines are still being created, the issue remains unresolved.

### Tracking information

```shell
REPLACE_ME
```

### Code to reproduce issue

```
No code is required, happens when the Tracking Server is started.
```

### Stack trace

```
No error
```

### Other info / logs

```
2024/08/20 16:54:57 INFO mlflow.store.db.utils: Create SQLAlchemy engine with pool options {'pool_size': 5, 'max_overflow': 5, 'pool_recycle': 1800, 'poolclass': <class 'sqlalchemy.pool.impl.QueuePool'>}
2024/08/20 16:54:58 INFO mlflow.store.db.utils: Create SQLAlchemy engine with pool options {'pool_size': 5, 'max_overflow': 5, 'pool_recycle': 1800, 'poolclass': <class 'sqlalchemy.pool.impl.QueuePool'>}
[2024-08-20 16:54:58 +0000] [14] [INFO] Starting gunicorn 22.0.0
[2024-08-20 16:54:58 +0000] [14] [INFO] Listening at: http://127.0.0.1:5000 (14)
[2024-08-20 16:54:58 +0000] [14] [INFO] Using worker: sync
[2024-08-20 16:54:58 +0000] [15] [INFO] Booting worker with pid: 15
[2024-08-20 16:54:58 +0000] [16] [INFO] Booting worker with pid: 16
[2024-08-20 16:54:58 +0000] [17] [INFO] Booting worker with pid: 17
[2024-08-20 16:54:58 +0000] [18] [INFO] Booting worker with pid: 18
127.0.0.1 - - [20/Aug/2024:16:56:04 +0000] "GET /mlflow/mlflow-test-3-3916cf/ HTTP/1.1" 200 0 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36"
2024/08/20 16:56:07 INFO mlflow.store.db.utils: Create SQLAlchemy engine with pool options {'pool_size': 5, 'max_overflow': 5, 'pool_recycle': 1800, 'poolclass': <class 'sqlalchemy.pool.impl.QueuePool'>}
2024/08/20 16:56:08 INFO mlflow.store.db.utils: Create SQLAlchemy engine with pool options {'pool_size': 5, 'max_overflow': 5, 'pool_recycle': 1800, 'poolclass': <class 'sqlalchemy.pool.impl.QueuePool'>}
2024/08/20 16:56:08 INFO mlflow.store.db.utils: Create SQLAlchemy engine with pool options {'pool_size': 5, 'max_overflow': 5, 'pool_recycle': 1800, 'poolclass': <class 'sqlalchemy.pool.impl.QueuePool'>}
2024/08/20 16:56:25 INFO mlflow.store.db.utils: Create SQLAlchemy engine with pool options {'pool_size': 5, 'max_overflow': 5, 'pool_recycle': 1800, 'poolclass': <class 'sqlalchemy.pool.impl.QueuePool'>}
2024/08/20 16:57:12 INFO mlflow.store.db.utils: Create SQLAlchemy engine with pool options {'pool_size': 5, 'max_overflow': 5, 'pool_recycle': 1800, 'poolclass': <class 'sqlalchemy.pool.impl.QueuePool'>}
2024/08/20 16:57:17 INFO mlflow.store.db.utils: Create SQLAlchemy engine with pool options {'pool_size': 5, 'max_overflow': 5, 'pool_recycle': 1800, 'poolclass': <class 'sqlalchemy.pool.impl.QueuePool'>}
2024/08/20 16:57:18 INFO mlflow.store.db.utils: Create SQLAlchemy engine with pool options {'pool_size': 5, 'max_overflow': 5, 'pool_recycle': 1800, 'poolclass': <class 'sqlalchemy.pool.impl.QueuePool'>}
2024/08/20 16:57:28 INFO mlflow.store.db.utils: Create SQLAlchemy engine with pool options {'pool_size': 5, 'max_overflow': 5, 'pool_recycle': 1800, 'poolclass': <class 'sqlalchemy.pool.impl.QueuePool'>}
```

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by harupy** [comment]
@alparslanavci Does this also happen with `--workers=1`?

**Comment by neo-edimar-cardoso** [comment]
> [@alparslanavci](https://github.com/alparslanavci) Does this also happen with `--workers=1`?

Yes, it happens even if you set 1 worker, in my case I'm using --uvicorn-opts and K8S HPA, so when my service scale to 4 pods I start to receive the error below and service outage due to hit the max connection limit:

sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) FATAL:  remaining connection slots are reserved for non-replication superuser, rds_reserved, and rds_superuser connections

I tried to limit my connection using pool_size and max_overflow, but it seems like this is being ignored, doesn't matter how many connections I configure, it always reaches the max connections under heavy load. _**MLFlow version used 3.5.1**_
